### PR TITLE
Satoshi's Calculator.

### LINKED ISSUE #62: Create Integration tests for bdk-cli
After Proof of Reserves capability is merged in https://github.com/bitcoindevkit/bdk/commit/b2ac4a0dfd33e1286059d0bd090fd9137e2545ba, its time we require some integration tests in the library. Not only for Proof of Reserves functionality, but in general wallet functions too.

This then can be added into CI tests to give us better reliability on the underlying bdk functions.

[electrsd](https://github.com/RCasatta/electrsd) can be used similarly to the bdk test framework to simulate a electrum and bitcoin core backend. https://github.com/bitcoindevkit/bdk/blob/64e88f0e006c68315142d53dc35b633327dde4b5/src/testutils/blockchain_tests.rs#L19-L22

### LINKED ISSUE #76: Open up rpc control for auto deployed regtest mode
The work needs to be done in two parts: 
 - Create non tmp datadir for regtest node data in `~/.bdk-bitcoin`.
 - Open up basic rpc calls via repl commands to do basic node operations.

### PR DESCRIPTION
**Pull request by rajarshimaitra**
### Description

fixes #62 
fixes #76 

On the name : 
If Bitcoin wallet devs had to have a "calculator", this is what it might look like. One single interface to do all wallet and node operations.  

This PR does the following
 - Update electrsd to latest version. We need this to use it for out auto deployed backends.
 - Add the remaining codes to `nodes.rs` module to define basic node operation apis on the `Nodes` enum.
 - Add a new `node` command in ` CliSubCommand` which has its own subcommand `NodeSubCommand`, These subcommands are essentially `bitcoin-cli` calls. And only the basic ones are included so far. We can also compose multiple `bitcoin-cli` calls to create our own commands in future.
 - Update the `handlers` to handle newly added `NodeSubCommands`.
 - Update in `utils` : 
   - This includes handling auto blockchain client creation for `regtest-*` features.
   -  update on the directory creation workflow. We are creating many directories for many stuffs, and I was trying to figure out the easiest way to keep the whole datadir struct intact, no matter where the user points its `datadir` to. Note that now `datadir` is a global app option, not just available for `regtest` nodes. Default `datadir` is the previous `~/.bdk-bitcoin`. 
   - Finally add an integration test in `tests/integration.rs` that runs a basic wallet operation with a bdk-cli wallet connected to auto deployed regtest-bitcoin/electrum.  

### Notes to the reviewers

@sandipndev @krtk6160. I feel this PR is now ready to try out #87 .

Also looking for more integration test ideas too add into.

basic `node` usage looks like this
```
$ ./target/debug/bdk-cli node --help
bdk-cli-node 0.5.0
Regtest Node mode

USAGE:
    bdk-cli node <SUBCOMMAND>

FLAGS:
    -h, --help       
            Prints help information

    -V, --version    
            Prints version information


SUBCOMMANDS:
    generate         Generate blocks
    getbalance       Get Wallet balance
    getinfo          Get info
    getnewaddress    Get new address from node's test wallet
    help             Prints this message or the help of the given subcommand(s)
    sendtoaddress    Send to an external wallet address
```
The biggest benefit is I think in the `repl` mode. That now kinda becomes an integrated tool the like python interpreter to operate a bdk-cli and a bitcoin backend from one single interface. 

I am excited to see what kind of demonstrations with bdk-cli we can create with this..  

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk-cli/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

### CODE REVIEW

**Comment by rajarshimaitra** [comment]
Pushed some more updates and refactoring.. Rearranged the commit history itself for easier reviews instead of new commits..

**Comment by rajarshimaitra** [comment]
@notmandatory I feel like #104 could also be included in this PR itself.. That has no code changes but documentation updates only..

**Comment by waterst0ne** [comment]
Hi @rajarshimaitra 
I tested out all the backend node subcommands `generate` , `getbalance`, `getinfo`, `getnewaddress`,and `sendtoaddress`. Everything seems to be operating smoothly on its own. 

After playing with it for a few hours I noticed a few things:
- when running `repl` on the help menu it is not clear to which commands belong to to `backend-node wallet` vs the `bdk-cli wallet` 
- when running `repl`  the `generate 100` subcommand has conflicting issues with the `bdk-cli generate` subcommand.  I suggest changing the `node generate` subcommand to `node generateblock` 

- This is a preview of what the help menu on repl looks like right now.
- A new dev would not know which commands are for backend or the bdk wallet. 
```
    generate             Generate blocks
    get_balance          Returns the current wallet balance
    get_new_address      Generates a new external address
    getbalance           Get Wallet balance for Node
    getinfo              Get info
    getnewaddress        Get new address from node's test wallet
```

-  I suggest changing the description to something that differentiates `backend node` vs `bdk-cli wallet` commands to something like this. 
```
    generate             [backend-node] Generate blocks
    get_balance          Returns the current wallet balance
    get_new_address      Generates a new external address
    getbalance           [backend-node] Get Wallet balance for Node
    getinfo              [backend-node]Get info
    getnewaddress        [backend-node] Get new address from node's test wallet

```

Overall everything looks good, Satoshi's Calculator is pretty awesome for learning I enjoyed using it Thanks!

File: src/nodes.rs
Comment (waterst0ne) [inline src/nodes.rs:72]:
```suggestion
            NodeSubCommand::Generateblock { block_num } => {
```

File: src/commands.rs
Comment (waterst0ne) [inline src/commands.rs:152]:
```suggestion
    Generateblocks { block_num: u64 },
```

**Comment by rajarshimaitra** [comment]
Thanks for the look @waterst0ne .. Those are good suggestions, will update them soon..

**Comment by rajarshimaitra** [comment]
Thanks @waterst0ne for the review.. Those were helpful.. I have addressed your comments on last 3 commits.

https://github.com/bitcoindevkit/bdk-cli/pull/102/commits/e728577bbf79cbc277abfee9f6e37264d9330408 - Instead of changing the command name, I elaborated the help doc. The command not only generate blocks but also funds the internal wallet.

https://github.com/bitcoindevkit/bdk-cli/pull/102/commits/c1921a7093c06b139ffb7e5eaaa3815885a903dc - Thanks for trying out the rel mode, and I agree having all the commands together is confusing.. I restructured the repl command and have them categorized in `node`, `wallet` and `key` subcommand.

So repl now looks like this
```
$ ./target/debug/bdk-cli repl --descriptor "wpkh(tprv8ZgxMBicQKsPeuazF16EdPZw84eHj55AU8ZKgZgdhu3sXcHnFgjzskfDvZdTaAFHYNCbKqrurFo9onSaT7zGT1i3u3j7LKhVZF5sJA39WPN/86'/1'/0'/0/*)#40hv8z77"
>> help
bdk-cli 0.5.0

USAGE:
    bdk-cli <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    exit      Exit REPL loop
    help      Prints this message or the help of the given subcommand(s)
    key       Execute Key Commands
    node      Execute Node Commands
    wallet    Execute wallet Commands

>> 
```

And for each subcommand you can ask for `help`.
```
>> node help
node 0.5.0
Execute Node Commands

USAGE:
    node <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    generate         Generate given number of blocks and fund the internal wallet with coinbases
    getbalance       Get Wallet balance
    getinfo          Get info
    getnewaddress    Get new address from node's test wallet
    help             Prints this message or the help of the given subcommand(s)
    sendtoaddress    Send to an external wallet address
```
https://github.com/bitcoindevkit/bdk-cli/pull/102/commits/e6e602c0d85a5dca96c1ea7fc805a0c1f3c76a8b - Adds all the `bitcoin-cli` rpc commands into our `node` subcommand.. So this now works too..
```
>> node listwallets
[
  "default"
]
>> node getbalances
{
  "mine": {
    "immature": 150.0,
    "trusted": 0.0,
    "untrusted_pending": 0.0
  }
}
>> 
```
Even if we don't have these commands listed. 

Things to figure out
 - How to convey this to bdk-cli users?
 - How to propagate `bitcoin-cli` command help to bdk-cli`.
 - Is having all the bitcoin-cli` commands available through bdk-cli useful?

**Comment by rajarshimaitra** [comment]
@notmandatory Should I also add #104 into this PR and get it all merged together??

**Comment by notmandatory** [comment]
> @notmandatory I feel like #104 could also be included in this PR itself.. That has no code changes but documentation updates only..

Looks like #104 is touching alot of files and already has review comments so let's keep it separate. 

Also I love the sub-command change you did for the repl, it makes the subcommands alot more organized and easier to understand what they're for.

**Comment by rajarshimaitra** [comment]
Yes makes sense.. Let me know if you find any more issue trying it out that we can push into this PR.. To me it seems now ready to merge.. There are few small updates I have in mind, but we can do them later via Issues and PRs.. 

But getting this and #104 in would complete the major refactorings.. 

 I would rebase #104 again once this gets merged..

**Comment by rajarshimaitra** [comment]
Also now this Fixes #110

**Comment by rajarshimaitra** [comment]
Updated with bdk v0.20.0 upgrade..

**Review by notmandatory — CHANGES_REQUESTED** [review]
I think this one is almost ready.. I just found a small issue for the CHANGELOG that should be fixed and then I can give it a utACK. 

@waterst0ne did you get a chance to redo the manual tests we did on this before these repl submenu changes?

**Comment by rajarshimaitra** [comment]
@waterst0ne reported a problem that I wanna document here.. 

bdk-cli doesn't seem to be working with this descriptor 
`wpkh(tprv8ZgxMBicQKsPd5F8dLQA1Wn5a9gcu5Dqt1RKSU1Arw3dTc7qmZPY7wzTbdiygXFE9yFJeUekPmxGXn35fQmmkSNebyXb3cJ86sbe32ACgNz/84'/0'/0'/*)`. 

I don't know how this was generated but this seems like a very valid descriptor like any other.. Here's the full parse of it by bdk..
```
wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/0'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: a066e93a, child_number: Hardened { index: 0 }, public_key: PublicKey(c2b157c5a54d046359f17e1c0b2f1a4b1bcc86e6ec6cf95a425585c0364a33c646bcbc849e81fabb4dbd9c8beb1ac07a8109e09559504186b38285697879357a), chain_code: 8329f5f6b2ccf23975f2b28a0802778ce27147a8b071cb68c44bfeb50edcb57a }, derivation_path: m, wildcard: Unhardened }))
```

Where as when I am creating a fresh descriptor from bdk-cli itself, its working all fine.. Heres an example descriptor generated by bdk-cli `"wpkh([cad5ef5e/84'/1'/0'/0]tprv8i78DGMvrqUS15nRQbQd3JtBCL4SsmGMMczTBQhgPgbEGjtnP5boKNidLv3HjkAzdr4FHjH4NVto9ngv2NGVGmSXBqnkiS5d4FSnpi3LWyz/*)"`

And here's the parse of it 
```
wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/1'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: 6b9efee5, child_number: Hardened { index: 0 }, public_key: PublicKey(ffefcdf5298f982aa396bae04b4cbffd13c0c8c85bac0d93cc3f129dcfed97f2e1a353f99329ac610057572d55e14cea765345e19f3e8915371da1efa2cdb617), chain_code: f831fa15022f13eb6fd92b025e54ceb0a4202bf75e3463ac4462c4bee15b764e }, derivation_path: m/0, wildcard: Unhardened }))
```

I am still not sure whats going on. Things that I have checked for the problematic descriptor
 - The bdk database doesn't update anything after the sync.
 - The corresponding bitcoin core wallet for that descriptor doesn't get any balance either. 
 
Is there any such thing as "rouge descriptors" that behaves in weird ways??

**Comment by waterst0ne** [comment]
> @waterst0ne reported a problem that I wanna document here..
> 
> bdk-cli doesn't seem to be working with this descriptor `wpkh(tprv8ZgxMBicQKsPd5F8dLQA1Wn5a9gcu5Dqt1RKSU1Arw3dTc7qmZPY7wzTbdiygXFE9yFJeUekPmxGXn35fQmmkSNebyXb3cJ86sbe32ACgNz/84'/0'/0'/*)`.
> 
> I don't know how this was generated but this seems like a very valid descriptor like any other.. Here's the full parse of it by bdk..
> 
> ```
> wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/0'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: a066e93a, child_number: Hardened { index: 0 }, public_key: PublicKey(c2b157c5a54d046359f17e1c0b2f1a4b1bcc86e6ec6cf95a425585c0364a33c646bcbc849e81fabb4dbd9c8beb1ac07a8109e09559504186b38285697879357a), chain_code: 8329f5f6b2ccf23975f2b28a0802778ce27147a8b071cb68c44bfeb50edcb57a }, derivation_path: m, wildcard: Unhardened }))
> ```
> 
> Where as when I am creating a fresh descriptor from bdk-cli itself, its working all fine.. Heres an example descriptor generated by bdk-cli `"wpkh([cad5ef5e/84'/1'/0'/0]tprv8i78DGMvrqUS15nRQbQd3JtBCL4SsmGMMczTBQhgPgbEGjtnP5boKNidLv3HjkAzdr4FHjH4NVto9ngv2NGVGmSXBqnkiS5d4FSnpi3LWyz/*)"`
> 
> And here's the parse of it
> 
> ```
> wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/1'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: 6b9efee5, child_number: Hardened { index: 0 }, public_key: PublicKey(ffefcdf5298f982aa396bae04b4cbffd13c0c8c85bac0d93cc3f129dcfed97f2e1a353f99329ac610057572d55e14cea765345e19f3e8915371da1efa2cdb617), chain_code: f831fa15022f13eb6fd92b025e54ceb0a4202bf75e3463ac4462c4bee15b764e }, derivation_path: m/0, wildcard: Unhardened }))
> ```
> 
> I am still not sure whats going on. Things that I have checked for the problematic descriptor
> 
>     * The bdk database doesn't update anything after the sync.
> 
>     * The corresponding bitcoin core wallet for that descriptor doesn't get any balance either.
> 
> 
> Is there any such thing as "rouge descriptors" that behaves in weird ways??

My balance did not update once again on the bdk-cli wallet side when I used `sendtoaddress` command.
Syncing issues.
I deleted the .bdk-bitcoin folder, sent the funds againn to the `bdk-cli wallet`, generated a few more  blocks using the node and still no balance. I'm not sure why I have this sync issue on regtest. I will continue to communicate  with Raj to see the reason for this case.

**Comment by waterst0ne** [comment]
> > @waterst0ne reported a problem that I wanna document here..
> > bdk-cli doesn't seem to be working with this descriptor `wpkh(tprv8ZgxMBicQKsPd5F8dLQA1Wn5a9gcu5Dqt1RKSU1Arw3dTc7qmZPY7wzTbdiygXFE9yFJeUekPmxGXn35fQmmkSNebyXb3cJ86sbe32ACgNz/84'/0'/0'/*)`.
> > I don't know how this was generated but this seems like a very valid descriptor like any other.. Here's the full parse of it by bdk..
> > ```
> > wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/0'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: a066e93a, child_number: Hardened { index: 0 }, public_key: PublicKey(c2b157c5a54d046359f17e1c0b2f1a4b1bcc86e6ec6cf95a425585c0364a33c646bcbc849e81fabb4dbd9c8beb1ac07a8109e09559504186b38285697879357a), chain_code: 8329f5f6b2ccf23975f2b28a0802778ce27147a8b071cb68c44bfeb50edcb57a }, derivation_path: m, wildcard: Unhardened }))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where as when I am creating a fresh descriptor from bdk-cli itself, its working all fine.. Heres an example descriptor generated by bdk-cli `"wpkh([cad5ef5e/84'/1'/0'/0]tprv8i78DGMvrqUS15nRQbQd3JtBCL4SsmGMMczTBQhgPgbEGjtnP5boKNidLv3HjkAzdr4FHjH4NVto9ngv2NGVGmSXBqnkiS5d4FSnpi3LWyz/*)"`
> > And here's the parse of it
> > ```
> > wpkh(XPub(DescriptorXKey { origin: Some((b62f919d, m/84'/1'/0')), xkey: ExtendedPubKey { network: Testnet, depth: 3, parent_fingerprint: 6b9efee5, child_number: Hardened { index: 0 }, public_key: PublicKey(ffefcdf5298f982aa396bae04b4cbffd13c0c8c85bac0d93cc3f129dcfed97f2e1a353f99329ac610057572d55e14cea765345e19f3e8915371da1efa2cdb617), chain_code: f831fa15022f13eb6fd92b025e54ceb0a4202bf75e3463ac4462c4bee15b764e }, derivation_path: m/0, wildcard: Unhardened }))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am still not sure whats going on. Things that I have checked for the problematic descriptor
> > ```
> > * The bdk database doesn't update anything after the sync.
> > 
> > * The corresponding bitcoin core wallet for that descriptor doesn't get any balance either.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is there any such thing as "rouge descriptors" that behaves in weird ways??
> 
> My balance did not update once again on the bdk-cli wallet side when I used `sendtoaddress` command. Syncing issues. I deleted the .bdk-bitcoin folder, sent the funds againn to the `bdk-cli wallet`, generated a few more blocks using the node and still no balance. I'm not sure why I have this sync issue on regtest. I will continue to communicate with Raj to see the reason for this case.

Turns out that when I send funds to the bdk-cli wallet it doesn't sync and balance doesn't update wether im on repl or not. Same results, not sure why. It was working before.

**Comment by rajarshimaitra** [comment]
Thanks to @waterst0ne we figured out this is a problem with current bdk v0.20.. I think this needs to wait for the next bdk release where the rpc problem is fixed..

**Comment by notmandatory** [comment]
Since [bdk 0.21](https://github.com/bitcoindevkit/bdk/releases/tag/v0.21.0) is out now can this PR be fixed? I think the PR that was needed is https://github.com/bitcoindevkit/bdk/pull/672.

**Comment by notmandatory** [comment]
I added a comment on #726.

File: src/commands.rs
Comment (danielabrozzoni) [inline src/commands.rs:1505]: Why did this change from bdk 0.19 to 0.20? o.o

Comment (notmandatory) [reply src/commands.rs:1505]: PR was started back in June, `bdk` and related dependencies will need to be updated to `0.22` once that's out.

File: src/commands.rs
Comment (rajarshimaitra) [reply src/commands.rs:53]: Good question.. While I have never thought about this before, but I think running bdk-cli with memory db doesn't make much sense. 

Each bdk-cli call starts a wallet, does operations, then drops the wallet. This approach only works when the wallet state is persisted in a db and the new wallet is loaded from the db, in subsequent calls, identified by the descriptor checksum value (unless use gives a name, then they need call bdk-cli methods with the same wallet name).

In case of memory db every call will be a fresh new wallet.. So you can't do much with it.

File: src/commands.rs
Comment (rajarshimaitra) [reply src/commands.rs:1505]: I actually don't remember fully.. But I think something was failing in the update from 19.0 to 20.0 and I did some quick changes to fix the test..

Comment (danielabrozzoni) [reply src/commands.rs:1505]: Cool :)

File: src/commands.rs
Comment (danielabrozzoni) [reply src/commands.rs:53]: OHHHHH you're right. Which means that #114 might not make sense.

But... I'm not sure if I can use the other dbs on WASM. I'll think more about this, if this PR is already merged when I find a solution I'll just update the code myself :)

Comment (notmandatory) [reply src/commands.rs:53]: Oh drat! I didn't see this before merging #114. 😬 It sounds like we may still need memory db support for wasm, or could also be used with the cli in repl mode.  But if you want me to revert #114 let me know.

Comment (notmandatory) [reply src/commands.rs:53]: I verified using the memory db with `repl` mode:
```shell
cargo run --no-default-features --features esplora-ureq,repl -- repl --descriptor "wpkh(tpubEBr4i6yk5nf5DAaJpsi9N2pPYBeJ7fZ5Z9rmN4977iYLCGco1VyjB9tvvuvYtfZzjD5A8igzgw3HeWeeKFmanHYqksqZXYXGsw5zjnj7KM9/*)"
```
It works fine, but with a couple build warnings:
```shell
warning: unused variable: `database_path`
   --> src/utils.rs:169:9
    |
169 |     let database_path = prepare_wallet_db_dir(wallet_name)?;
    |         ^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_database_path`
    |
    = note: `#[warn(unused_variables)]` on by default

warning: variable does not need to be mutable
   --> src/utils.rs:131:9
    |
131 |     let mut db_dir = prepare_home_wallet_dir(wallet_name)?;
    |         ----^^^^^^
    |         |
    |         help: remove this `mut`
    |
    = note: `#[warn(unused_mut)]` on by default
```

Comment (rajarshimaitra) [reply src/commands.rs:53]: I am very confused.. Did the wallet work worked in repl mode?? Like, could you send sats to it, sync and get the balance?? 

I think there might be few ways to make it only work in `repl` mode. But as the current code stands, it should not work.. Or I am missing something important..

Comment (rajarshimaitra) [reply src/commands.rs:53]: The warnings happen because the code is written with a "always db" as assumption.. If we need memory db for wasm we might need to modify the code a bit, along with the build script.. Getting bdk-cli into wasm should be an important enough task to do whatever is needed..

Comment (notmandatory) [reply src/commands.rs:53]: > I am very confused.. Did the wallet work worked in repl mode with memory db?? Like, could you send sats to it, sync and get the balance?? It might not throw any error, but it won't be functional wallet if wallet data gets wiped out after every call..

Yes the cli wallet works with a memory db in repl mode using the above feature settings. I verified syncing, receiving sats from a testnet faucet, list_unspent, get_balance. The reason it works is because in repl mode the wallet (and memory db) is created before you enter the repl loop.

Comment (rajarshimaitra) [reply src/commands.rs:53]: Ahh right.. Its not the command where the iteration happens, its the loop.. Thanks.. Then it works.. #114 seems good to me.. So we just need to clear up few things to make room for memory db in the code.. Also should we only allow `repl` mode while using memory db?

Comment (notmandatory) [reply src/commands.rs:53]: I don't think we need to limit using the mem db to only repl mode. I can imagine a wasm web wallet that isn't using the repl feature, but does work like repl with a single Wallet instance that it sync's on startup and keeps live until page is closed. We already make sqlite-db a default feature so most casual users shouldn't accidentally enabled the memory db.

Comment (rajarshimaitra) [reply src/commands.rs:53]: Okay.. Ya if that works then it would be sweet.. I will rebase this after #115 . I think that one is clean to go too.. And update to get memory db without the warnings..

**Comment by rajarshimaitra** [comment]
@notmandatory added a v0.22 update patch here https://github.com/bitcoindevkit/bdk-cli/pull/102/commits/716efe8681be22b55a386516c979cf5b5e6cbe5a pointing to the `bdk-reserve` open PR https://github.com/bitcoindevkit/bdk-reserves/pull/11.. 

But for some reason the git link is not working in CI for bdk-reserve. It seems to work fine in my local.. 


Did the v0.22.0 update here instead of master, because we have many modification over master here, and wanted to cover them through.

Not Much changes in the code due to update. But we might need to change the `RpcOpts` as many new options are added in `RpcSyncParam` in bdk v0.22.. `skip_block` is not used in RPC anymore, so we might remove that from cli options.. But all these can be done later in separate PR.

**Comment by notmandatory** [comment]
prelim ACK, @waterst0ne and I did testing on earlier versions in July of this and the main suggestion from that testing was addressed when @rajarshimaitra added sub-menus in the repl. I'd like to go through it again one more time once #118 is merged and this PR re-based on it. Plus we need the next bdk-reserves release including https://github.com/bitcoindevkit/bdk-reserves/pull/11.  Long story short, I think this PR is only missing dependency updates.

**Comment by rajarshimaitra** [comment]
Rebased on master and consolidated commit history..

**Comment by notmandatory** [comment]
Everything looks good with the electrum backend and bdk client, I did a bunch of testing with repl mode too. I pushed a small commit so that in repl mode the backend node is only created once, otherwise it's too slow to use. I also found that with rpc mode the bdk wallet is unable to sync, in the integration test or from the cli. The error I'm getting is below. I think this is likely related to [this issue](https://github.com/casey/ord/issues/273) in the ord project.  Should we just comment out the `regtest-bitcoin` feature for now? I have a feeling this rpc client issue isn't going to be a quick fix.

```
RUST_LOG=info cargo run --features rpc -- --network regtest wallet --descriptor "wpkh(tprv8ZgxMBicQKsPdusu4AAEfp6mxEfdoLHjvPnNwKVSr6A4XdqWE8vTEcxuJE2AsdFuscZDuSLfU8PEwC5aBJBTS4wMZhnC4Esy5bodQAGFRQ5/*)" --cookie "/Users/steve/.bdk-bitcoin/bitcoind/regtest/.cookie" --node "127.0.0.1:59848" -s 10000000000 sync
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/bdk-cli --network regtest wallet --descriptor 'wpkh(tprv8ZgxMBicQKsPdusu4AAEfp6mxEfdoLHjvPnNwKVSr6A4XdqWE8vTEcxuJE2AsdFuscZDuSLfU8PEwC5aBJBTS4wMZhnC4Esy5bodQAGFRQ5/*)' --cookie /Users/steve/.bdk-bitcoin/bitcoind/regtest/.cookie --node '127.0.0.1:59848' -s 10000000000 sync`
[2022-09-10T02:18:23Z INFO  bdk::database::sqlite] db up to date, no migration needed
[2022-09-10T02:18:23Z INFO  bdk::blockchain::rpc] connected to 'http://127.0.0.1:59848/wallet/l32feux5' with auth: Cookie { file: "/Users/steve/.bdk-bitcoin/bitcoind/regtest/.cookie" }
[2022-09-10T02:18:23Z INFO  bdk::blockchain::rpc] wallet already loaded: l32feux5
[2022-09-10T02:18:23Z INFO  bdk::blockchain::rpc] initial db state: txs=0 utxos=0
[2022-09-10T02:18:38Z ERROR bdk_cli] Rpc(JsonRpc(Transport(SocketError(Os { code: 35, kind: WouldBlock, message: "Resource temporarily unavailable" }))))
```

**Comment by rajarshimaitra** [comment]
Thanks @notmandatory for the report. Unfortunately I am not being able to create this in local. But the error message looks very familiar as Evan reported in this issue. https://github.com/bitcoindevkit/bdk/issues/650#issuecomment-1182990519

It probably means there's an bitcoind instance already running that is accessing the node data directory, thus the `Resource temporarily unavailable` error.. Can you check that all the bitcoind processes are being cleared after each run?

Also from the command
```
RUST_LOG=info cargo run --features rpc -- --network regtest wallet --descriptor "wpkh(tprv8ZgxMBicQKsPdusu4AAEfp6mxEfdoLHjvPnNwKVSr6A4XdqWE8vTEcxuJE2AsdFuscZDuSLfU8PEwC5aBJBTS4wMZhnC4Esy5bodQAGFRQ5/*)" --cookie "/Users/steve/.bdk-bitcoin/bitcoind/regtest/.cookie" --node "127.0.0.1:59848" -s 10000000000 sync
```
It seems you are running regular `rpc` mode.. Not the `regtest-bitcoin` mode.. A command like this working both normal as well as repl mode for me
```
cargo run --features regtest-bitcoin -- --network regtest wallet --descriptor "wpkh(tprv8ZgxMBicQKsPdusu4AAEfp6mxEfdoLHjvPnNwKVSr6A4XdqWE8vTEcxuJE2AsdFuscZDuSLfU8PEwC5aBJBTS4wMZhnC4Esy5bodQAGFRQ5/*)" sync
```
 


I checked your changes too.. ACK on taking the backend creation to global.. Though `new_backend()` probably doesn't need feature gate, as it can run for all features, and for non `regtes-*` feature it just won't do anything.. But this can be done later.. 

I tested the work flow again in linux machine and all seems to be working okay..

**Comment by notmandatory** [comment]
> I tested the work flow again in linux machine and all seems to be working okay..

Thanks for retesting on Linux, the problem I'm seeing must just be on MacOS. And for ACKing my global  repl `new_backend()` change, I had to put the feature gate so CI clippy doesn't give a warning when no backend feature is enabled.

**Review by notmandatory — APPROVED** [review]
tACK fa0454e9728ceea5bdbffb9ed0e51a32db7c74ab

**Pull request merged**
